### ISSUE TITLE
STAR runs out of memory during samtools sort step

### ISSUE THREAD

**Issue by rdmorin** [LCR-BCCRC/lcr-modules#94]
When testing the demo on Cedar, the STAR step fails when running samtools sort because it runs out of memory. This is after already increasing the default value to 24000. I suspect this has not been run on Numbers either because presumably the same problem would happen on any Slurm cluster with this data set. Any suggestions for how to determine what a reasonable default should be? 

```
srun: error: cdr510: task 0: Exited with exit code 1
srun: Terminating job step 49869634.0
/cvmfs/soft.computecanada.ca/nix/var/nix/profiles/16.09/bin/bash: line 1:  6557 Killed                  samtools sort -@ 12 -m $((24000 / 12))M -T results/star-1.3/02-sort_bam/mrna--grch37/TCRBOA7-T-RNA -o results/star-1.3/02-sort_bam/mrna--grch37/TCRBOA7-T-RNA.sort.bam results/star-1.3/02-sort_bam/mrna--grch37/TCRBOA7-T-RNA.bam > results/star-1.3/logs/launched-2020-09-03-at-12-02-18/02-sort_bam/mrna--grch37/TCRBOA7-T-RNA/bam_sort.stdout.log 2> results/star-1.3/logs/launched-2020-09-03-at-12-02-18/02-sort_bam/mrna--grch37/TCRBOA7-T-RNA/bam_sort.stderr.log
```

**Comment by rdmorin** [comment]
Even setting this value to 48G doesn't solve the problem. There must be something else going on here.

**Comment by Kdreval** [comment]
I did run the test with our demo sample on Numbers before pushing the 1.3 version. Also, I used Numbers for the latest test on August 30 with demo_repairs branch. I did not have memory errors (or other errors with STAR) during these runs

**Comment by lkhilton** [comment]
I ran into this issue with Ferran's RNAseq data on Numbers, but only for one of four samples, and increasing the memory to 40 GB was sufficient.

**Comment by lkhilton** [comment]
I would wager this has to do with how the memory is allocated per thread in the utils module. It's set to divide the allocated memory equally across the number of threads, which means there's no buffer if Samtools briefly exceeds the allocated memory for any reason. I ran into this issue with GRIDSS and had to set my JVM heap size to 80% of the allocated memory for all tools. We should change this part of the utils_sort_bam rule: 

```
-m $(({resources.mem_mb} * 0.8 / {threads}))
```

**Comment by rdmorin** [comment]
I suspected the same. My change was a more aggressive reduction from the value specified and it worked but I'll try this instead.

**Comment by lkhilton** [comment]
If you want to know how much memory it actually uses and whether your buffer will be enough, after the jobs complete you can use this command: 

```
sacct -u $USER --format="JobID,JobName,AllocCPUS,CPUTime,MaxRSS,MaxVMSize,ReqMem,ExitCode,Elapsed,Start"
```

MaxRSS is the key value to know how much total memory the job used.

**Comment by rdmorin** [comment]
Thanks. It is fairly quick to die if it is going to die so I'll probably just re-run it after the markdups step completes. Did you ever run this module on Numbers? If so, I don't understand why this didn't come up sooner.

**Comment by lkhilton** [comment]
Yeah, I ran it on Numbers for the fusion bake-off with Ferran's data. Only one of the four samples errored out, and increasing the memory in the config fixed the problem, so I wasn't motivated at the time to dive deeper into this

**Comment by rdmorin** [comment]
Markdups has also been killed. Can someone explain what exactly this does?

resources:
        mem_mb = CFG["mem_mb"].get("utils_bam_markdups", 8000)

**Comment by lkhilton** [comment]
It sets a default value if mem_mb isn't set in the config

**Comment by rdmorin** [comment]
```
/bin/bash: 24000 * 0.5 / 12: syntax error: invalid arithmetic operator (error token is ".5 / 12")
```
@lkhilton any idea why this worked for you but something similar didn't work for me? If I change this to 24000 / 12 / 2 it works

**Comment by rdmorin** [comment]
I resolved this but I'm not totally happy with having to divide the memory request in half before passing it to samtools. I'm also curious why the above didn't work.

**Comment by lkhilton** [comment]
What was the resolution to this, Ryan? Did you make changes to the utils module that will go on master to deal with the problem of the files exceeding allotted memory?

**Comment by Kdreval** [comment]
I had just had this same problem yesterday. I do not know what was causing it, but you earlier in this thread suggested that Samtools briefly exceeds the allocated memory and this causes for scheduler to kill the process, and I think this might be the case. I was working on demo project for picard branch, and addressed this issue in [commit](https://github.com/LCR-BCCRC/lcr-modules/commit/47e9008b87bb0c0b762c39a9610b223459160e42) that was merged to master today.

**Comment by lkhilton** [comment]
Thanks, Kostia. I'm running the version that was on master previously and having many problems. I'm testing with 80% of the allocated memory divided per thread. We should think about making this standard across lcr-modules, I'm sure we will one day come across other tools that have this same problem.

**Comment by Kdreval** [comment]
The most recent merge from today's morning merge have demo that I run to completion yesterday. I did it on numbers and used the wrapper from demo project folder. This latest merge might help resolve the problems you have. I agree some other tools might need this adjustment, too. Even for the STAR, I did it on the demo sample only, but other samples may behave differently and it will still need adjustment. Let me know if I can help with demo, I think I am the latest person to run it to completion.
